[#10] 비동기 쿠폰 발급 기능 구현

codesejin opened the pull request:
관련 Issue : #10 

# 쿠폰 발급 비동기 API
![image](https://github.com/f-lab-edu/off-coupon/assets/101460733/ddbbcb89-cb13-47de-b97c-2bb6eeaadfdf)

- [x] Redis를 활용한 요구사항 검증
       - Redis를 이용하여 쿠폰 발급 요청에 대한 데이터를 저장하고 관리합니다.
       - 쿠폰 발급 요청에 대한 유효성을 Redis를 통해 검증하여 신뢰성을 확보합니다.
- [x] RabbitMQ를 활용한 대기큐에 메세지 적재
       - RabbitMQ를 통해 쿠폰 발급 요청 메시지를 대기큐에 적재합니다.
       - 대기큐에 적재된 메시지는 비동기적으로 처리됩니다.
- [x] 스케줄링을 활용해 N개 트랜잭션 작업을 분산
       - 스케줄러를 활용하여 대기큐에서 적재된 메시지를 정기적으로 확인하고, 쿠폰 발급 처리를 일정 간격으로 진행합니다.
       - 이를 통해 시스템의 안정성을 유지하고 트래픽 부하를 분산합니다.
- [x] 메세지 consume할때마다 SSE 알림 전송
       - 쿠폰 발급 이력에 대한 SSE 알림을 구현하여, 클라이언트에게 실시간으로 쿠폰 발급 상태를 알립니다.
       - 각 쿠폰 발급 이력이 처리될 때마다 해당 이력에 대한 SSE 알림을 전송합니다.

![비동기 알람](https://github.com/f-lab-edu/off-coupon/assets/101460733/669ee4f2-78c2-4b1c-82ee-1a6ebb785d09)

## SSE 구현 시 주의 사항 (동시성 이슈)

- SSE는 클라이언트가 서버에 연결한 후 서버에서 지속적으로 데이터를 푸시하는 방식
- 여러 클라이언트가 동시에 연결되면 서버는 각 클라이언트에 대해 개별적인 연결을 유지해야 한다.
- 해결 : ConcurrentHashMap에 SSE객체 담기

 ---
 
# TODO
1. RabbitMQ를 Junit으로 테스트 가능한지 알아봐야합니다.
2. 테스트 코드 작성 미흡해서 추가해야합니다.
3. SecurityContextHolder에서 계속 로그인한 상태에서 Anonymous User로 나오는 원인을 파악하지 못했습니다.

# 질문
CouponIssueConsumer클래스에서 인스턴스 변수로 사용한 CouponIssueMessageForQueue 가 분산 서버 환경에서 여러 coupon일 경우 겹치지 않을지 잘 모르겠습니다.


---

# 변경된 SSE 설계

- ConcurrentHashMap만 사용하면 여러 WAS에서 Session이든 memberId든 식별자 공유 못하는 현상 발생
- SseEmitter를 Redis에 저장하려고 했으나, 통신을 담당하고 있어서 Serialize가 불가능
- 따라서 Redis의 Pub/Sub기능을 활용해서 토픽(memberId 기준)을 구독하고 있는 subscriber에게 RabbitMQ메세지가 도착하면 SSE알림 전송으로 변경
<img width="867" alt="image" src="https://github.com/f-lab-edu/off-coupon/assets/101460733/7121bc97-b259-49a2-9766-21c5fa447fb0">

codesejin committed:
test : 초기화시 생성사 주입 수정

codesejin committed:
feature : RabbitMq, Schedule 비동기 API, SSE 연결

codesejin committed:
refactor : 주석 추가 및 리팩토링

codesejin committed:
refactor : 필요없는 테스트 삭제

codesejin committed:
refactor : 소나클라우드 냄새나는 코드 개선

codesejin committed:
refactor : html 화면 출력 수정

f-lab-maverick commented on load-test/docker-compose.yml:
docker compose를 적용하려고 하신건가요? 어떤것을 적용하려고 시도하신건가요?

codesejin replied:
locus라는 부하테스트 툴을 통해 동기랑 비동기 API 테스트하려고 했습니다!
Jmeter에서는 RPS가 나오지 않아서 다른 성능 지표도 확인해보고자 했습니다

- TPS(Transaction Per Second) : 1초당 처리할 수 있는 트랜잭션의 수
- RPS(Request Per Second) : 1초에 처리하는 HTTP 요청 수

f-lab-maverick commented on src/main/java/com/flab/offcoupon/config/security/configs/SecurityConfig.java:
```diff
@@ -27,24 +29,37 @@ public class SecurityConfig {
+                                        "/",
+                                        "/api/v1/members/signup",
+                                        "/api/v1/event/**", // TODO : 테스트용
+                                        "/api/v1/sse/**",
```
테스트용으로 열어둔 권한은 잊지말고 이후에 꼭 권한처리해주세요~

codesejin replied:
넵 수정했습니답..!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/controller/HomeController.java:
```diff
@@ -1,15 +1,23 @@
+import org.springframework.web.servlet.ModelAndView;
 
+@Slf4j
 @RequestMapping("/")
 @RestController
 public class HomeController {
 
-    @GetMapping
+    @GetMapping("/")
     public String home() {
         return "home";
     }
+
+    @GetMapping("/main")
```
`Controller` 레벨에 `RequestMapping`으로 `/`를,
메소드 레벨에 추가적인 매핑을 `/`, `/main`으로 해둔다면 중첩되어서 요청 URL이 `//`, `//main`이 됩니다.

Spring에서 이런 부분을 어느정도 처리해서 괜찮겠지만, 논리적으로 이슈가 생기는 부분이니 `Controller` 레벨에서 `RequestMapping`을 제거해주시는건 어떠실까요?

codesejin replied:
넵 수정했습니다 감사합니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/domain/sse/ConnectionPoolInterface.java:
커넥션풀과 관련된 로직을 추상화하려고 한 시도로 보이네요.
인터페이스로 분리하여서 어떤 장점을 가져가려고 하셨나요?

codesejin replied:
추후에 sse가 아닌 fcm으로 기능 변경이 변경될때를 대비해서 인터페이스로 분리했습니다.
그런데 fcm은 제가 직접 구현해보진 않았지만 메소드명에서 sse와 차이가 있을것 같습니다...

f-lab-maverick commented on src/main/java/com/flab/offcoupon/domain/sse/SseConnectionPool.java:
Socket 방식이 아닌 SSE를 선택한 이유가 있을 것으로 보입니다~ SSE와 소켓의 장단점을 비교해보셨을텐데 어떤 이유에서 SSE를 고르셨나요?
또한, 직접 폴링 / 롱폴링등의 기법을 사용해볼 수도 있었을 것으로 보이는데 어떤 기능들과 비교해보셨나요?

codesejin replied:
1. `직접 폴링` : 클라이언트가 특정 주기로 request를 날려서 이벤트를 받는 방식인데, http overhead가 많이 발생합니다. http 요청 커넥션을 맺고 끊는걸 반복하기 때문에 불필요한 네트워크 호출이 많이 발생합니다.
2. `롱 폴링` : 클라이언트가 서버로 request를 보내면 대기중이다가 서버에서 이벤트를 받으면 연결이 종료되는 방식입니다. 직접 폴링에 비해 네트워크 호출이 적기 때문에 http overhead가 적게 발생하지만, 다수의 클라이언트에게 동시 이벤트가 발생할 경우 서버 접속이 급격하기 늘어나면서 서버 부하가 발생합니다.
4. `socket` : 단일 tcp/ip를 사용하는 방식으로 클라이언트와 양방향으로 메세지를 주고받는것이 아니기 때문에 사용하지 않았습니다.
5. `sse` : 추가적으로 socket연결을 하지 않아도 기존의 http연결을 재사용해 데이터를 전송하기 때문에 별도의 프로토콜이 필요하지 않습니다. 그리고 서버에서 단방향으로 이벤트를 알려주는 것이기 때문에 sse를 사용했습니다.

이 외에도 `fcm`같은 경우에도 push 알림이 가능하고, 다양한 플랫폼에서 사용할 수 있지만 sse에 비해 실시간성이 느립니다. 이미 비동기 방식으로 일정한 시간동안 쿠폰 발급이 진행되기 때문에 여기서 더 지연시간이 발생한다면 사용자 경험이 떨어질것이라고 생각했습니다.

f-lab-maverick commented on src/main/java/com/flab/offcoupon/domain/sse/SseConnectionPool.java:
```diff
@@ -0,0 +1,41 @@
+    private static final Map<String, UserSseConnection> connectionPool = new ConcurrentHashMap<>();
+
+
+    @Override
+    public void addSession(String uniqueKey, UserSseConnection userSseConnection) {
+        connectionPool.put(uniqueKey, userSseConnection);
+    }
+
```
이미 Session이 있는 경우에 대한 처리는 어떻게되나요?

codesejin replied:
기본적으로 Map을 사용했기 때문에  key에 동일한 session이 있을 경우 value가 대체 됩니다.
그런데 value이 경우 userSseConnection 객체를 통해 session과 커넥션풀, json으로 변형하기 위한 objectmapper가 생성되므로 , 사실상 계속 동일 session을 사용하는거랑 다름이 없습니다

f-lab-maverick commented on src/main/java/com/flab/offcoupon/domain/sse/SseConnectionPool.java:
```diff
@@ -0,0 +1,41 @@
+        connectionPool.put(uniqueKey, userSseConnection);
+    }
+
+    @Override
+    public UserSseConnection getSession(String uniqueKey) {
+        return connectionPool.get(uniqueKey);
+    }
```
Session이 조회되지 않는 경우에는 어떻게 될까요?

codesejin replied:
Session이 조회되지 않을 경우 UserSseConnection는 null이 되고,
이상태로 메세지를 push할때 어떤 session으로 보낼지 값이 나오지 않습니다.


세션이 조회되지 않을 경우 exception 처리하는걸로 변경하겠습니다
![image](https://github.com/f-lab-edu/off-coupon/assets/101460733/491a7f0f-16d2-407e-8982-a99c10319bbf)

f-lab-maverick commented on src/main/java/com/flab/offcoupon/domain/sse/SseConnectionPool.java:
```diff
@@ -0,0 +1,41 @@
+     * 각각 다른 스레드에서 연결 정보를 저장, 꺼내기, 발송하기 작업이 수행됩니다.
+     * 이로 인해 스레드 간의 경합이 발생할 수 있으며, 이를 해결하기 위해 ConcurrentHashMap을 활용했습니다.
+     */
+    private static final Map<String, UserSseConnection> connectionPool = new ConcurrentHashMap<>();
```
유저의 세션을 특정 서버에 보관한다면, 서버가 다운되었을 때 장애상황에 대처할 수 있는 방법이 있을까요?

codesejin replied:
단일 서버에서 세션을 보관하는데 다운될 경우 데이터 유실이 발생합니다.
그래서 저는 redis를 활용해 session storage방식을 사용했는데, redis를 master/slave 혹은 클러스터링으로 구성해서 복제하면 장애를 예방할 수 있습니다

f-lab-maverick replied:
현재 소스코드에서는 `private static final Map<String, UserSseConnection> connectionPool = new ConcurrentHashMap<>();` 형태로 서버의 메모리 내부에 저장하고 있습니다. 별도로 redis로 저장하는 로직이 없는데 유실에 대한 대비는 따로 없으신건가요?

codesejin replied:
아 부분은 redis역직렬화 문제로 트러블 이슈가 있어서, 특정서버에 보관하는 걸로 대체했습니답..

codesejin replied:
redis에 별도로 저장하려고 시도해봤는데 이렇게 한 이유를 설명드리자면,

1. Redis에 Object를 저장하려면 `ObjectMapper`를 사용해서 Json으로 변환을 해줘야됩니다.
2. Redis에 저장하려는 객체는 UserSseConnection 타입인데, UserSseConnection 내부에 또 다른 객체가 필드로
SseEmitter, SseConnectionPool, ObjectMapper 3가지가 있습니다.
3.  ObjectMapper를 사용해서 Json으로 변형하려면  타입으로 된 필드의 내부에도 getter가 있어야 가능합니다.
```
public class Jobs {
    private String jobName;
    private int salary;

    public String getJobName() {
        return jobName;
    }

    public int getSalary() {
        return salary;
    }
}
```
<img width="887" alt="image" src="https://github.com/f-lab-edu/off-coupon/assets/101460733/c0370d2c-0ab9-4bf5-91af-51864e9b6ff6">

4. 그런데  각 필드가 프레임워크또는 인터페이스 타입이라  getter사용이 불가능하고 @JsonIgnore를 사용하면  Redis에 timeout이 null되버려서 괜찮을까 고민했습니다.
```
{
  "uniqueKey": "1",
  "sseEmitter": {
    "timeout": 60000 // -> null로 됨
  },
  "connectionPoolIfs": {},  // -> null로 됨
  "objectMapper": {}  // -> null로 됨}
```
```userSseConnection = UserSseConnection(uniqueKey=1, sseEmitter=null, connectionPoolIfs=null, objectMapper=null)```

6. null로되면 redis에서 반환받을때 UserSseConnection내부에 있는 메소드 기능을 활용하기가 어렵습니답


<img width="822" alt="image" src="https://github.com/f-lab-edu/off-coupon/assets/101460733/9cce3e96-a815-4847-a547-94b0e1e1932a">


이부분은 객체 설계 관련해서 고민해봐야할 문제 같아요. 그런데 또 말씀해주신 분산환경 상에서 특정 서버에만 적용되는거라 설계부분을 어떻게할지 고민중입니당


추가적으로 로직을 다시 생각해봤을때 message queue에서 받은 memberId로 알림을 전송하는거라 session이 아닌 memberId로 식별해야 할것 같다고 판단이 되어 메소드명을 수정했습니답

f-lab-maverick replied:
Redis에 저장하는 DTO와 실제 비즈니스 DTO를 분리하면 해결될 이슈로 보이네요. 오늘 좀 더 자세히 살펴보며 개선해야할 점을 좀 더 피드백드리겠습니다.

1. `UserSseConnection`은 불필요한 파라미터인 `objectMapper`를 들고있습니다. 메세지를 직렬화하고 역직렬화하려는 의도로 보이는데, 이건 커넥션의 역할이 아니죠. 메세지는 <String>으로 통일하고 직렬화는 이 커넥션을 이용하는 서비스가 직접 직렬화하여 입력하도록 만들면 불필요한 변수를 줄일 수 있습니다.
2. 이 서비스에서 사용자에게 SSE를 위해 저장해야 하는 것은 사용자의 유니크 키 정도입니다. 이 커넥션을 미리 만들어놔야할 이유가 있나요? 사용자의 유니크 id정도만 따로 관리하고 이 커넥션객체는 SSE를 보낼 때 생성해서 사용해도 되지 않을까요?

codesejin replied:
말씀해주신대로 변경해보겠습니다 감사합니다!

또 다른 이슈사항이 있어서, 설계를 수정했습니답

-----
- 기존 설계 : RabbitMQ에서 메세지를 consume하면 바로 SSE 알람 전송 
-  변경된 설계 : RabbitMQ에서 메세지를 consume하면 Redis의 Publisher가 memberId기준으로 토픽 생성 후 메세지 전송
- SSE를 연결하고 있는 쪽에서는 Redis Subscriber로서 해당 토픽을 구독하고 있다가 메세지를 받으면 SSE 알람 전송
- 이유 : 기존 설계는 여러 was에서 sseemitter객체를 공유할 수 없었음(ConcurrentHashMap사용
- 그래서 Redis에 memberId와 SseEmitter객체를 저장하려고 했으나, SseEmitter는 통신을 담당하기 때문에 Serialize해도 재사용 못함
- 결국 Redis Pub/Sub을 사용해서 토픽을 구독중인 유저에게 SSE 알람 전송하는 방식으로 수정

f-lab-maverick commented on src/main/java/com/flab/offcoupon/service/coupon_issue/async/consumer/CouponIssueConsumer.java:
```diff
@@ -0,0 +1,114 @@
+     *
+     * @throws IOException 메시지 큐 조회 중 발생하는 예외
+     */
+    @Transactional
+    @Scheduled(fixedDelay = 3000)
+    public void consumeCouponIssueMessage() throws IOException {
+        if (existCouponIssueQueueTarget()) {
+            message = (CouponIssueMessageForQueue) rabbitTemplate.receiveAndConvert(QUEUE_NAME);
+            log.info("'coupon-issue.queue'에 메시지가 있습니다. message: {}", message);
+            saveEachCouponIssueHistory(message);
+            sseAlertSender.pushSseMessage(message.memberId(),"쿠폰이 발급 완료되었습니다. memberId : %s");
+        } else {
+            log.info("'coupon-issue.queue'가 비어있습니다.");
+            totalUpdateIssuedCouponAndDeleteRequest();
+        }
+    }
+    /**
```
Rabbit MQ에서 데이터를 들고오는것부터 트랜잭션이 시작할 이유가 있을까요?

트랜잭션의 범위는 최소한으로 줄여야 undo영역에 데이터가 많이 쌓이는 것을 방지할 수 있고, 성능과 데이터이슈에도 유리합니다. 또한 Queue에 메세지가 없는데 매번 트랜잭션을 시작할 필요도 없겠네요~

Queue에서 데이터를 들고오는 부분은 트랜잭션과 연결될 이유가 없으니 트랜잭션 외부에서 진행하도록 해주세요~

codesejin replied:
메세지큐를 사용하지 않을때도 스케줄링때문에 3초마다 트랜잭션이 열릴것이라고 생각을못했네요!
알려주셔서 감사합니다! 수정했습니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/service/coupon_issue/async/consumer/CouponIssueConsumer.java:
```diff
@@ -0,0 +1,114 @@
+     */
+    @Transactional
+    @Scheduled(fixedDelay = 3000)
+    public void consumeCouponIssueMessage() throws IOException {
```
`IOException`과 같은 checked exception을 메소드 외부로 throw하게 되면 꽤나 이슈가 있습니다.
1. IOException이 발생할 경우 트랜잭션이 롤백되는 조건에 해당할까요?
2. 이 메서드를 사용하는 모든 메서드에서 `try-catch`를 사용해야 하지 않을까요?

이러한 이슈가 있기 때문에 Exception wrapping & Rethrowing 기법을 많이 사용합니다. 이 키워드로 한번 찾아보세요~

codesejin replied:
1. @Transactional 사용 시 아무런 옵션없이 사용한다면 기본적으로 Unchecked Exception과 Errors만 확인 후 롤백 시킵니다. IOException은 Checked Exception이기 때문에 트랜잭션 내부에 Checked Exception이 발생할 여지가 있다면 `@Transactional(rollbackFor = Exception.class)` 를 설정해줘야 합니다.
-> 리팩토링과정에서 트랜잭션을 분리했기 때문에, 트랜잭션 범위내에 IOException이 존재하지 않아서 별도로 설정하지 않겠습니다.

3. 말씀해주신것 처럼 처음 메소드에서 throws를 사용(`예외 회피`)하다보니 해당 메소드를 호출하는 상위메소드에도 예외 전파 현상이 발생합니다. 

CheckedException은 복구 불가능이기 때문에 최초 메소드에서  Exception catch -> UncheckedException으로
`Exception wrapping`을 통해 `예외 전환`을 사용하여 Exception Hadling을 하도록 변경했습니다.
<img width="913" alt="image" src="https://github.com/f-lab-edu/off-coupon/assets/101460733/c37c389b-38d8-4e31-bce2-6e98fe46cd2c">

f-lab-maverick commented on src/main/java/com/flab/offcoupon/util/CouponRabbitMQUtils.java:
```diff
@@ -0,0 +1,28 @@
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+
+/**
+ * RabbitMQ 에서 사용하는 exchange, queue, routing key 를 생성하는 유틸리티 클래스
+ */
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public class CouponRabbitMQUtils {
+    /**
+     * RabbitMQ에서 사용하는 Exchange의 이름입니다.
+     * 메시지를 받고 적절한 Queue로 라우팅하는 역할을 담당합니다.
+     */
+    public static final String EXCHANGE_NAME = "coupon-issue.exchange";
```
이 클래스틑 아직까지는 상수만 담아두는 역할을 하고있네요. 별도 유틸 메서드가 없다면 네이밍을 상수 클래스라는걸 알 수 있게 변경해주세요.

추가적으로, 유틸클래스의 경우 생성자를 없애는것도 좋지만 lombok에서 제공해주는 [`@UtilityClass`](https://projectlombok.org/features/experimental/UtilityClass)를 사용하실 수도 있습니다. 꽤 여러가지 기능을 해주니 한번 문서를 읽어보세요~

codesejin replied:
@UtilityClass를 사용하면,  컴파일 시점에 생성자 private 으로 만들어 주고, 모든 메서드를 static으로 만들어 줘서 더 명시적인것 같습니다 알려주셔서 감사합니다!

상수만 있는 클래스는 Utils대신 Constants로 변경하겠습니다!

f-lab-maverick commented on src/main/resources/application.properties:
```diff
@@ -23,6 +23,18 @@ spring.data.redis.host=${REDIS_HOST}
 spring.data.redis.password=${REDIS_PASSWORD}
 
+# RabbitMQ
+spring.rabbitmq.host=${RABBITMQ_HOST}
+spring.rabbitmq.port=${RABBITMQ_PORT}
+spring.rabbitmq.username=${RABBITMQ_USERNAME}
+spring.rabbitmq.password=${RABBITMQ_PASSWORD}
```
비밀정도 숨기기 좋습니다 👍

codesejin replied:
감사합니다!

f-lab-maverick commented on src/test/java/com/flab/offcoupon/repository/MemberRepositoryTest.java:
```diff
@@ -21,22 +21,29 @@
+
     @BeforeEach
     void setUp() {
+        setupUtils = new SetupUtils();
```
이렇게 인스턴스화를 하는 대상은 유틸 클래스라고 보기는 어렵네요.
유틸리티성 클래스의 경우 상태가 없어야 합니다.

특정 값을 세팅해주는게 목표라면 `Initializer`, `Factory` 등의 네이밍을 사용해보시는건 어떨까요?

codesejin replied:
Initializer로 변경하겠습니다 감사합니다!

f-lab-maverick reviewed (changes requested):
고생하셨습니다~ 리뷰 확인 후 다시 요청해주세요!

codesejin committed:
refactor : 스케줄링으로 부터 트랜잭션 분리 & @UtilityClass적용

codesejin committed:
refactor : 클래스명, 예외 핸들링 수정
1. 상수만 있다면 클래스명은 Utils 말고 상수임을 나타낼 수 있도록 변경
2. 예외 전환 및 예외 래핑

codesejin committed:
refactor : 사용하지 않는 메소드 제거

codesejin committed:
refactor : 주석 및 인가 수정

codesejin reviewed (commented):
리뷰해주신 내용대로 수정했습니다..!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/service/coupon_issue/async/consumer/CouponIssueConsumer.java:
```diff
@@ -0,0 +1,65 @@
+     * 쿠폰 발급 메세지를 저장하기 위한 변수
+     * 스케줄링이 매번 호출되기 때문에 Queue가 비어있을 경우 couponId를 찾기 위해 static으로 선언했습니다.
+     */
+    public static CouponIssueMessageForQueue message;
```
3초마다 스케줄링이 호출된다면, 만약 로직이 3초를 넘어서는 시간동안 실행된다면 실행되는 동안 이 `message`변수가 다른 변수로 재할당 되며 버그가 날 위험성이 있습니다.

또한 `public static` 접근제어자이기 떄문에 다른 클래스에서 이 변수를 가져다쓰다가 다시 가져다쓰려고 할 때 또다른 인스턴스로 재할당되어서 다른 인스턴스가 튀어나오기도 하겠네요.

외부에서 아래처럼 이 변수를 null로 재할당하는 로직도 있어서 상당히 위험한 로직으로 보입니다.
```
totalUpdateIssuedCouponAndDeleteRequest
```


이렇게 message 변수를 참조하며 상태를 변경하는 메서드들이 정말 동시다발적으로 많이 발생한다면 굉장히 큰 버그로 다가올 수 있습니다. 동시성 이슈가 꽤나 많이 일어날 수 있는 영역이며, 별도의 동시성 제어또한 들어가있지 않아 꼭 수정이 필요한 영역으로 보입니다.

해결방법으로 애초에 public static 변수가 필요한지에 대해 고려해보면 좋을거같네요.
필요할 때 지역변수로 생성하여 만들어두고, 이후 폐기하면 되지 않을까요?


아래 메서드를 
```
@Scheduled(fixedDelay = 3000)
    public void consumeCouponIssueMessage() {
        if (existCouponIssueQueueTarget()) {
            message = (CouponIssueMessageForQueue) rabbitTemplate.receiveAndConvert(QUEUE_NAME);
            log.info("'coupon-issue.queue'에 메시지가 있습니다. message: {}", message);
            couponIssueMessageHandler.saveEachCouponIssueHistory(message);
            sseAlertSender.pushSseMessage(message.memberId(),"쿠폰이 발급 완료되었습니다. memberId : %s");
        } else {
            log.info("'coupon-issue.queue'가 비어있습니다.");
            couponIssueMessageHandler.totalUpdateIssuedCouponAndDeleteRequest(message);
        }
    }
```

이렇게 바꿀 수 있습니다.
```
@Scheduled(fixedDelay = 3000)
    public void consumeCouponIssueMessage() {
        if (existCouponIssueQueueTarget()) {
            CouponIssueMessageForQueue  message = (CouponIssueMessageForQueue) rabbitTemplate.receiveAndConvert(QUEUE_NAME);
            log.info("'coupon-issue.queue'에 메시지가 있습니다. message: {}", message);
            couponIssueMessageHandler.saveEachCouponIssueHistory(message);
            sseAlertSender.pushSseMessage(message.memberId(),"쿠폰이 발급 완료되었습니다. memberId : %s");
        } else {
            log.info("'coupon-issue.queue'가 비어있습니다.");
            couponIssueMessageHandler.totalUpdateIssuedCouponAndDeleteRequest(message);
        }
    }
```

codesejin replied:
멘토링 시간때 말씀해주신대로 스케줄링 2번하는걸로 변경했습니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/util/DateTimeUtils.java:
```diff
@@ -1,13 +1,12 @@
 import java.time.ZonedDateTime;
 
-@NoArgsConstructor (access = AccessLevel.PRIVATE)
+@UtilityClass
```
이 어노테이션을 붙이면 메서드에서 static 키워드를 빼도 static이 붙여진것으로 인식되어 동작합니다.
다만 가독성이나 헷갈리지 않는 영역 때문에라도 static키워드는 그대로 남겨두는게 더 좋아보이네요 :)

f-lab-maverick reviewed (changes requested):
확인하였습니다~ 리뷰 확인해주시고 다시 요청해주세요.

codesejin committed:
refactor : static 변수 제거 후 새로운 스케줄링 추가

DB부하를 줄이기 위함이었지만, static변수는 동시성 문제가 있기 때문에 스케줄링에서 사용되어도 절대 사용하지말아야 한다고 하셨다.
그래서 새로운 트랜잭션과 DB access가 늘어났는데 최적화 된 쿼리인걸까?

codesejin reviewed (commented):
sse관련해서 세션을 concurrentHashMap에다 저장했는데 objectmapper 로 역직렬화과정에서 트러블 이슈가 있어서 하지못했습니다. 아래 코멘트에 추가적으로 달았습니다!

codesejin reviewed (commented):
PR리뷰 주신대로 반영했는데, Redis에 sse관련해서 Redis에 저장하는 부분은  ObjectMapper 에서 역직렬화할때 고려해야할 상황이 있습니답

codesejin committed:
style : 필요없는 주석 지우기

codesejin committed:
refactor : 메소드명 수정

codesejin committed:
style : Lock 사용한 이유 주석 추가

f-lab-maverick commented on src/main/java/com/flab/offcoupon/component/sse/SseAlertSender.java:
```diff
@@ -21,8 +21,8 @@ public class SseAlertSender {
-    public void pushSseMessage(Long memberId, String message) {
-        UserSseConnection userSseConnection = sseConnectionPool.getSession(String.valueOf(memberId));
+    public void pushSseMessage(long memberId, String message) {
+        UserSseConnection userSseConnection = sseConnectionPool.getUniqueKey(String.valueOf(memberId));
```
👍

f-lab-maverick reviewed (commented):
고생하셨습니다. 간단한 코멘트 남겨두었으니 확인해주세요~

codesejin committed:
refactor : sse연결 설계 수정

기존 설계 : RabbitMQ에서 메세지를 consume하면 바로 SSE 알람 전송
변경된 설계 : RabbitMQ에서 메세지를 consume하면 Redis의 Publisher가 memberId기준으로 토픽 생성 후 메세지 전송
-> SSE를 연결하고 있는 쪽에서는 Redis Subscriber로서 해당 토픽을 구독하고 있다가 메세지를 받으면 SSE 알람 전송
이유 : 기존 설계는 여러 was에서 sseemitter객체를 공유할 수 없었음(ConcurrentHashMap사용)
-> 그래서 Redis에 memberId와 SseEmitter객체를 저장하려고 했으나, SseEmitter는 통신을 담당하기 때문에 Serialize불가
-> 결국 Redis Pub/Sub을 사용해서 토픽을 구독중인 유저에게 SSE 알람 전송하는 방식으로 수정

codesejin reviewed (commented):
수정 완료했습니다~

codesejin committed:
refactor : 예외 회피 -> 예외 전환으로 수정

codesejin committed:
refactor : 스프링 시큐리티 주석 일단 제외

codesejin committed:
refactor : 테스트용 println제거

codesejin committed:
refactor : Exception Wrapping

f-lab-maverick commented on src/main/java/com/flab/offcoupon/config/ObjectMapperConfig.java:
```diff
@@ -0,0 +1,13 @@
+
+@Configuration
+public class ObjectMapperConfig {
+    @Bean
+    public ObjectMapper objectMapper() {
+        return new ObjectMapper();
+    }
```
Spring Web MVC모듈을 쓰신다면 `ObjectMapper`는 이미 bean으로 등록이 되어 있습니다~ 별도로 등록하지 않아도 됩니다.

codesejin replied:
알려주셔서 감사합니다!

f-lab-maverick reviewed (approved):
고생하셨습니다~ 머지해주셔도 좋습니다.

codesejin committed:
refactor : ObjectMapper는 Spring Web MVC모듈에서 이미 Bean으로 등록되어 있음
